### PR TITLE
[Draft] PON 2.0

### LINKED ISSUE #1: Allow non-letter characters in variable names
If I understand the parser correctly variable names can only be upper or lowercase letters.
In JSON every string is possible as name. It would be nice if this is supported aswell. Maybe you can add `<<>>` brackets if it contains other characters.

Also to create any string there would need to be an escape sequence.
Maybe the escape character could be `<`? It is not intuitive and already used as start of a string.

### PR DESCRIPTION
**Pull request by sertonix**
So these would be my ideas for a larger update.
Some comments would be nice!

### Goals

1. Lossless conversion from and to JSON (excluding comments).
2. Make it a bit more awfull.

### Changes

| status| description
|---|---
| done! | type definitions define the complete path to the value and not only the last key. |
| done! | keys can be quoted like strings to allow any character. |
| writer only | indent with tab and 3 spaces |
| done! | add Array type |
| done! | remove useful error messages |
| missing | force multiple lines for multiline comments |
| done! | add string escape sequence |
| done! | non object root element |

(+ some minor changes and major simplifications)

- [ ] TODO write specs 2.0

fix #1

### CODE REVIEW

**Comment by confused-Techie** [comment]
I love the effort here, and will review this as soon as I can. Feel free to ping over on Discord to talk further

**Comment by confused-Techie** [comment]
Thanks for adding Mauricios ideas for improving our error handling here.

One quick question, since I'm hoping to give this repo some attention today, do you think we should include all your previous PRs into the V2.0 of PON? I kinda feel like that makes sense, and then prior to merging them we can tag the current repo as V1.

What do you think?

**Comment by confused-Techie** [comment]
Oh actually maybe scratch that. It looks like this is what's already being done? By including things like the CLI rework in here, does this also address your previous PRs?

**Comment by sertonix** [comment]
I have merged all my other pull requests into this one.
So all my PRs and issues can be closed with this.

**Comment by confused-Techie** [comment]
@Sertonix Okay awesome that makes sense.

And any ideas like adding a few other error messages should they be PRs to this, or do you mind if I commit directly

File: lib/pon-parser.pegjs [hunk 1]
```diff
@@ -1,33 +1,129 @@
+{
+  const Types = require("./types.js");
+  
+  function pathToString(path) {
```
Comment (confused-Techie) [inline lib/pon-parser.pegjs:4]: Is there a reason we have `pathToString` doubled up here?

**Comment by confused-Techie** [comment]
Also fyi I'll take a look at getting indentation to work on read

File: lib/pon-parser.pegjs (hunk 1 above)
Comment (sertonix) [reply lib/pon-parser.pegjs:4]: No. Fixed that

**Comment by sertonix** [comment]
I think this pull request should be ready to merge.
Finishing the rest (like indentation) can be done in further pull requests.

**Comment by confused-Techie** [comment]
Ahh yeah fair point actually @Sertonix 

In that case lemme tag what the repo has then we can merge this one

**Review by confused-Techie — APPROVED** [review]
Alright the changes here are beautiful, I'll go ahead and merge this one, and start a doc for the new spec if you wanna help get that fleshed out. Like I see the new usage of a `Gimme` symbol right? 

But yeah once fleshed out we can get the syntax highlighting updated as well

**Comment by sertonix** [comment]
The main spec changes are:
- definitions include the whole path
- strings (not the type actually) can include parts in binary (and comments) inside `<>`
- type names, keys (the parts of a path), values are all strings or unquoted names
- `Gimme` can be used to use the given key as return value

**Pull request merged**